### PR TITLE
#4252: Build tt-metal with clang-17 instead of gcc

### LINKED ISSUE #4252: Enable C++20
@eyonland @arakhmati and myself can see big benefits with adapting C++20. This can be a contained rollout as we can compile a subset of our code with C++20, and the rest still on C++17. (See https://stackoverflow.com/questions/46746878/is-it-safe-to-link-c17-c14-and-c11-objects/49119902#49119902). 

However, we need to migrate to at least GCC 10 as many required [features](https://gcc.gnu.org/projects/cxx-status.html) are only present in the newer versions. Luckily, Ubuntu 20.04 has a GCC 10 [package](https://packages.ubuntu.com/search?keywords=gcc-10). This won't be a problem if/when we migrate to conda as I've already used conda's GCC 12 package.

### PR DESCRIPTION
**Pull request by ttmchiou**
Needs https://github.com/tenstorrent-metal/tt-metal/pull/6835 merged first.

PR enables building tt-metal with Clang-17

### CODE REVIEW

**Comment by tt-rkim** [comment]
Are we merging this first, or?
We don't need to make any code changes to tt-metal to make this work?

**Comment by tt-rkim** [comment]
We probably should update the deps stuff, like in `.github/actions`

**Comment by ttmchiou** [comment]
@tt-rkim 
PR needs https://github.com/tenstorrent-metal/tt-metal/pull/6835 merged first.

Deps are already installed in all CI VMs, and the provisioning already includes clang-17 installation.
soon to be Dev VMs as well.
Just pre-emptively getting the PRs ready. pending https://github.com/tenstorrent-metal/tt-metal/pull/6835

**Comment by tt-rkim** [comment]
Can you try running the eager package main then, to confirm that compilation and stuff works?

**Comment by tt-rkim** [comment]
C++ build binaries job would be good here too

**Comment by tt-rkim** [comment]
Once other one merged in

File: .github/actions/install-metal-dev-deps/action.yml [hunk 1]
```diff
@@ -46,3 +46,10 @@ runs:
         cmake -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_SHARED_LIBS=ON .
         make
         sudo make install
+    - name: Install clang-17
+      shell: bash
+      run: |
+        wget https://apt.llvm.org/llvm.sh
+        chmod u+x llvm.sh
+        sudo ./llvm.sh 17
+        clang-17 --version
```
Comment (tt-rkim) [inline .github/actions/install-metal-dev-deps/action.yml:55]: POSIX line

**Comment by tt-rkim** [comment]
```
2024-03-27T19:56:39.1589961Z make[1]: Leaving directory '/tmp/build-via-sdist-wegedgus/metal_libs-0.46.0rc3.dev26+wormhole.b0/tt_metal/hw'
2024-03-27T19:57:52.3684082Z In file included from tt_metal/impl/dispatch/command_queue.cpp:5:
2024-03-27T19:57:52.3685297Z In file included from ./tt_metal/impl/dispatch/command_queue.hpp:15:
2024-03-27T19:57:52.3692205Z ./tt_metal/impl/dispatch/lock_free_queue.hpp:55:38: error: 'iterator<std::forward_iterator_tag, tt::tt_metal::CommandInterface>' is deprecated [-Werror,-Wdeprecated-declarations]
2024-03-27T19:57:52.3694908Z    55 |         class Iterator : public std::iterator<std::forward_iterator_tag, T> {
2024-03-27T19:57:52.3695799Z       |                                      ^
2024-03-27T19:57:52.3698035Z tt_metal/impl/dispatch/command_queue.cpp:1597:26: note: in instantiation of member class 'LockFreeQueue<tt::tt_metal::CommandInterface>::Iterator' requested here
2024-03-27T19:57:52.3699651Z  1597 |     for (const auto& cmd : tq.worker_queue) {
2024-03-27T19:57:52.3700324Z       |                          ^
2024-03-27T19:57:52.3705874Z /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_iterator_base_types.h:127:12: note: 'iterator<std::forward_iterator_tag, tt::tt_metal::CommandInterface>' has been explicitly marked deprecated here
2024-03-27T19:57:52.3712160Z   127 |     struct _GLIBCXX17_DEPRECATED iterator
2024-03-27T19:57:52.3712831Z       |            ^
2024-03-27T19:57:52.3715805Z /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13/bits/c++config.h:121:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
2024-03-27T19:57:52.3717397Z   121 | # define _GLIBCXX17_DEPRECATED [[__deprecated__]]
2024-03-27T19:57:52.3718116Z       |                                  ^
2024-03-27T19:57:55.3832002Z 1 error generated.
2024-03-27T19:57:55.4185133Z make: *** [/tmp/build-via-sdist-wegedgus/metal_libs-0.46.0rc3.dev26+wormhole.b0/tt_metal/impl/module.mk:35: /tmp/build-via-sdist-wegedgus/metal_libs-0.46.0rc3.dev26+wormhole.b0/build/obj/tt_metal/impl/dispatch/command_queue.o] Error 1
2024-03-27T19:57:55.4188200Z make: *** Waiting for unfinished jobs....
```

Iterator is deprecated? Wonder if it's the `CONFIG=release` here...

**Comment by ttmchiou** [comment]
> Can you try running the eager package main then, to confirm that compilation and stuff works?
> C++ build binaries job would be good here too

2024-03-28T22:11:10.9458149Z ./tt_metal/impl/dispatch/lock_free_queue.hpp:55:38: error: 'iterator<std::forward_iterator_tag, tt::tt_metal::CommandInterface>' is deprecated [-Werror,-Wdeprecated-declarations\

it's just because we have -Werror which treats deprecations as errors. the use of std::iterator is deprecated in C++17 and entirely removed in C++20

**Comment by ttmchiou** [comment]
Closing pull request as addressed in 
https://github.com/tenstorrent/tt-metal/pull/9070

**Pull request closed without merging**